txnsync/encodedgroupsunmarshalers.go

```diff
@@ -0,0 +1,925 @@
+// Copyright (C) 2019-2021 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package txnsync
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/transactions"
+)
+
+var errDataMissing = errors.New("failed to decode: data missing")
+
+func getSlice(b []byte, index int, size int) ([]byte, error) {
+	if index*size+size > len(b) {
+		return nil, errDataMissing
+	}
+	return b[index*size : index*size+size], nil
+}
+
+func getNibble(b []byte, index int) (byte, error) {
+	if index > len(b)*2 {
+		return 0, errDataMissing
+	}
+	if index%2 == 0 {
+		return b[index/2] / 16, nil
+	}
+	return b[index/2] % 16, nil
+}
+
+func (stub *txGroupsEncodingStub) reconstructSignedTransactions(signedTxns []transactions.SignedTxn, genesisID string, genesisHash crypto.Digest) (err error) {
+	var index int
+	index = 0
+	err = stub.BitmaskSig.Iterate(int(stub.TotalTransactionsCount), func(i int) error {
+		slice, err := getSlice(stub.Sig, index, len(crypto.Signature{}))
+		if err != nil {
+			return err
+		}
+		copy(signedTxns[i].Sig[:], slice)
+		index++
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	if err := stub.reconstructMsigs(signedTxns); err != nil {
+		return fmt.Errorf("failed to msigs: %w", err)
+	}
+	if err := stub.reconstructLsigs(signedTxns); err != nil {
+		return fmt.Errorf("failed to lsigs: %w", err)
+	}
+	index = 0
+	err = stub.BitmaskAuthAddr.Iterate(int(stub.TotalTransactionsCount), func(i int) error {
+		slice, err := getSlice(stub.AuthAddr, index, crypto.DigestSize)
+		if err != nil {
+			return err
+		}
+		copy(signedTxns[i].AuthAddr[:], slice)
+		index++
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	return stub.reconstructTransactions(signedTxns, genesisID, genesisHash)
+}
+
+func (stub *txGroupsEncodingStub) reconstructMsigs(signedTxns []transactions.SignedTxn) (err error) {
+	var index int
+	index = 0
+	err = stub.BitmaskVersion.Iterate(int(stub.TotalTransactionsCount), func(i int) error {
+		if index >= len(stub.Version) {
+			return errDataMissing
+		}
+		signedTxns[i].Msig.Version = stub.Version[index]
+		index++
+		return nil
+	})
```

Comment:
similar to the above, this clearly indicate that you're maintaining an index which is increased on each call to the anonymous function. I think that moving this index into the Iterate method, and passing this onto the anonymous function via a parameter would simplify the implementation.

Reply:
also, maybe pass the `len(stub.Version)` here as an argument to the `stub.BitmaskVersion.Iterate` so that the Iterate could perform the array length validation ?